engine/flink/executor/src/main/scala/pl/touk/nussknacker/engine/process/scenariotesting/DataRecordsSource.scala

```diff
@@ -0,0 +1,128 @@
+import pl.touk.nussknacker.engine.api.runtimecontext.{ContextIdGenerator, EngineRuntimeContext}
+import pl.touk.nussknacker.engine.flink.api.compat.ExplicitUidInOperatorsSupport
+import pl.touk.nussknacker.engine.flink.api.process._
+import pl.touk.nussknacker.engine.flink.api.typeinfo.option.OptionTypeInfo
+import pl.touk.nussknacker.engine.flink.api.typeinformation.TypeInformationDetection
+import pl.touk.nussknacker.engine.flink.typeinformation.ConcreteCaseClassTypeInfo
+import pl.touk.nussknacker.engine.flink.watermarkstrategy.FlinkWatermarkStrategyRuntimeHandler
+import pl.touk.nussknacker.engine.flink.watermarkstrategy.FlinkWatermarkStrategyRuntimeHandler.{
+  ContextWithEventTime,
+  EventTimeTimestampAssigner
+}
+import pl.touk.nussknacker.engine.util.watermarkstrategy.WatermarkStrategyOptions
+
+import java.time.Instant
+import scala.jdk.CollectionConverters._
+
+class DataRecordsSource(
+    dataRecords: List[DataRecord],
+    sourceOutputValidationContext: ValidationContext,
+    watermarkStrategyOptions: Option[WatermarkStrategyOptions],
+) extends FlinkSource
+    with BaseFlinkSource
+    with ExplicitUidInOperatorsSupport {
+
+  override final def contextStream(
+      env: StreamExecutionEnvironment,
+      flinkNodeContext: FlinkCustomNodeContext
+  ): DataStream[Context] = {
+    val streamOfRaw =
+      sourceWithUidAndName(fromDataProperHandlingEmptyList(env), flinkNodeContext)
+
+    streamOfRaw
+      .flatMap(
+        new DataRecordContextInitializingFunction(
+          flinkNodeContext.nodeId,
+          flinkNodeContext.convertToEngineRuntimeContext,
+          // We have to convert lazy parameter to serializable form because for purpose of source-specific format, all component parameters
+          // were converted to evaluable form - see TestSourceComponentImplementationInvoker.invokeOriginalInvoker
+          watermarkStrategyOptions.map(_.eventTimeLazyParam).map(_.toSerializable),
+          flinkNodeContext.lazyParameterHelper,
+        ),
+        FlinkWatermarkStrategyRuntimeHandler.contextInitializingFunctionOutputTypeInfo(
+          flinkNodeContext.asOneOutputContext
+        )
+      )
+      .assignTimestampsAndWatermarks(
+        watermarkStrategyOptions
+          .map(FlinkWatermarkStrategyRuntimeHandler.watermarkStrategy)
+          .getOrElse(
+            WatermarkStrategy
+              .forMonotonousTimestamps[ContextWithEventTime]
+              .withTimestampAssigner(EventTimeTimestampAssigner)
+          )
+      )
+      .map((ctxWithEventTime: ContextWithEventTime) => ctxWithEventTime.context, flinkNodeContext.contextTypeInfo)
+  }
+
+  // env.fromData uses DataGeneratorSource which even for an empty list try to process one element
+  // which ends up with NoSuchElementException in FromElementsGeneratorFunction.tryDeserialize
+  private def fromDataProperHandlingEmptyList(env: StreamExecutionEnvironment): DataStreamSource[DataRecord] = {
```

Comment:
What are cases where we want to build DataRecordsSource with empty list of records? I assume we can't just treat this as illeagal state. Is it for situations where you test a scenario with multiple sources but not all sources had records specified for them and we need to stub them?

Reply:
Yes, this is one of those situations. I also use it during the scenario state verification process, which basically behaves similarly to scenario testing, but with empty input test records and uses dumped scenario state as an initial scenario state.